### PR TITLE
#163518701 Bookmark an article to read it later

### PR DESCRIPTION
**Pull request by Musigwa-zz**
#### What does this PR do?
- It helps the logged in user to bookmark an article for reading it later
#### Description of Task to be completed?
- Make sure that the user will be able to bookmark an article
- If that article is bookmarked by him, the bookmark will be removed/deleted

#### How should this be manually tested?
- After cloning the repository, check out `ft-create-article-endpoints-163518689` branch
- Make sure to read `README.md` file to get the setup instructions
- Using any `API client tool(usually POSTMAN)`,
- Create the user and login to get the `access-token`
- Create an article using `POST /api/v1/articles` endpoint
- Make sure to pass the token under `request headers`. It should be {`Authorization`: `Bearer <access-token>`} KEY, VALUE PAIR
- The request's body should be as mentioned in the `README.md` file under `Endpoints:` ==> `Add Comments to an Article` subtitle
After the successful article creation,
- Bookmark that article using this endpoint ==> `POST /api/v1/articles/:slug/bookmark`. It will only require passing the {`Authorization`: `Bearer <access-token>`} KEY, VALUE PAIR in the request headers and passing the correct article's slug as mentioned above in request URL.

#### Any background context you want to provide?
I deleted `helpers/data.js` as it contained the testData that we already have in `tests/helpers/testData`
#### What are the relevant pivotal tracker stories?
- #163518701

### CODE REVIEW

File: tests/integration/comment.test.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            if (!err) {
+              const { token: accessToken } = res.body;
+              token = accessToken;
+              chai
+                .request(app)
+                .post("/api/v1/articles")
+                .set("Authorization", `Bearer ${token}`)
+                .send(newArticle)
+                .end((err, res) => {
+                  console.log("object");
```
Comment (YvesIraguha) [inline tests/integration/comment.test.js]: Please, remove this `console.log()`.

**Review by YvesIraguha — APPROVED** [review]

File: controllers/articlesController.js [hunk 2]
```diff
@@ -3,11 +3,18 @@ import constants from "../helpers/constants";
...
-const { User, Article, Tag, ArticleTag } = models;
-const { CREATED, NOT_FOUND, BAD_REQUEST } = constants.statusCode;
+const { User, Article, Tag, ArticleTag, Bookmark } = models;
+const {
+  CREATED,
+  NOT_FOUND,
+  INTERNAL_SERVER_ERROR,
+  BAD_REQUEST,
+  GONE
+} = constants.statusCode;
```
Comment (puritybirir) [inline controllers/articlesController.js:14]: comment JSDoc before class

File: controllers/articlesController.js [hunk 3]
```diff
@@ -44,11 +56,66 @@ export default class PostController {
...
+  /**
+   * @description This helps the user to bookmark the article for reading it later.
+   * @param  {object} req - The request object
+   * @param  {object} res - The response object
+   * @return {object} - It returns the request response object
+   */
+
+  static async bookmark(req, res) {
+    try {
+      // TODO: Fetch the use by req.user.id,
```
Comment (puritybirir) [inline controllers/articlesController.js]: Remove comments before making pull requests

File: controllers/commentController.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import Joi from "joi";
+import models from "../models";
+import constants from "../helpers/constants";
+
+const { User, Article, Comment } = models;
+const { CREATED, BAD_REQUEST } = constants.statusCode;
+
```
Comment (puritybirir) [inline controllers/commentController.js:7]: JSDoc comment before class

File: tests/integration/comment.test.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+          expect(res.status).equals(CREATED);
+          expect(res.body.message).to.contain("Comment created");
+          expect(res.body).to.haveOwnProperty("comment");
+          expect(res.body.comment).to.haveOwnProperty("createdAt");
+          expect(res.body.comment).to.haveOwnProperty("author");
+          expect(res.body.comment.author).to.haveOwnProperty("username");
+          expect(res.body.comment.author.username).to.contain(dummyUser.username);
+          done();
+        });
+    });
```
Comment (puritybirir) [inline tests/integration/comment.test.js]: Correct consistency in all the tests

File: tests/integration/comment.test.js (hunk 1 above)
Comment (Musigwa-zz) [reply tests/integration/comment.test.js]: I have applied the feedback

File: controllers/articlesController.js (hunk 2 above)
Comment (Musigwa-zz) [reply controllers/articlesController.js:14]: I have applied the feedback.

File: tests/integration/comment.test.js (hunk 5 above)
Comment (Musigwa-zz) [reply tests/integration/comment.test.js]: I have applied the feedback. Thanks

File: controllers/commentController.js (hunk 4 above)
Comment (Musigwa-zz) [reply controllers/commentController.js:7]: Thanks, I have applied the feedback.

File: tests/integration/article.test.js [hunk 6]
```diff
@@ -121,4 +123,61 @@ describe("# Articles endpoints", () => {
...
+    });
+
+    it("should decline bookmarking the article which doesn't exist", done => {
+      const { article } = createdArticle;
+      chai
+        .request(app)
+        .post(`/api/v1/articles/${article.slug}fs1/bookmark`)
+        .set("Authorization", `Bearer ${token}`)
+        .end((err, res) => {
+          expect(res.status).equals(BAD_REQUEST);
```
Comment (espoirMur) [inline tests/integration/article.test.js]: Is BAD_REQUEST the Correct error code here?

File: tests/integration/article.test.js [hunk 7]
```diff
@@ -121,4 +123,61 @@ describe("# Articles endpoints", () => {
...
+          expect(res.body.message).to.contain("The article with this slug");
```
Comment (espoirMur) [inline tests/integration/article.test.js:165]: Is this message meaningful for users?

**Review by espoirMur — CHANGES_REQUESTED** [review]

File: tests/integration/article.test.js (hunk 6 above)
Comment (Musigwa-zz) [reply tests/integration/article.test.js]: Thanks, I have applied the feedback.

**Review by espoirMur — APPROVED** [review]

**Pull request merged**